src/runtime/index.ts

```diff
@@ -0,0 +1,288 @@
+
+import { Context, createSignal, DEV, JSX } from "solid-js";
+import createProxy, { setComponentProperty } from "./create-proxy";
+import isListUpdated from "./is-list-updated";
+
+interface ComponentOptions {
+  location?: string;
+  // In granular mode. This signature is a hash
+  // generated from the component's JS string
+  signature?: string;
+  // An array of foreign bindings (values that aren't locally declared in the component)
+  dependencies?: any[];
+}
+
+// The registration data for the components
+export interface ComponentRegistrationData<P> extends ComponentOptions {
+  // A compile-time ID generated for the component, this is usually
+  // derived from the component's name
+  id: string;
+  // The component itself
+  component: (props: P) => JSX.Element;
+  // This function replaces the previous component
+  // with the new component.
+  update: (action: () => (props: P) => JSX.Element) => void;
+}
+
+// The registration data for the context
+export interface ContextRegistrationData<T> {
+  // A compile-time ID generated for the context, this is usually
+  // derived from the context's name
+  id: string;
+  // The context instance
+  context: Context<T>;
+}
+
+export interface Registry {
+  components: Map<string, ComponentRegistrationData<any>>;
+  contexts: Map<string, ContextRegistrationData<any>>;
+}
+
+export function $$registry(): Registry {
+  return {
+    components: new Map(),
+    contexts: new Map(),
+  };
+}
+
+export function $$component<P>(
+  registry: Registry,
+  id: string,
+  component: (props: P) => JSX.Element,
+  options: ComponentOptions = {},
+): (props: P) => JSX.Element {
+  if (!component.name) {
+    setComponentProperty(component, 'name', id);
+  }
+  if (options.location) {
+    setComponentProperty(component, 'location', options.location);
+  }
```

Comment:
Actually, I'm not sure I have a way to access this function? 🤔
The `owner.fn` of proxy component would be the proxy
And `owner.fn` of the hmr memo is [this](https://github.com/solidjs/solid-refresh/blob/c98cb9ad830dab8621d7cac1e61f3ef9d3f3ab59/src/runtime/create-proxy.ts#L32)
So wouldn't it be better to add the location/name to the `get` traps of the proxy component?

Reply:
@thetarnav I agree that `createComponent` should be modified for that purpose.